### PR TITLE
[inductor] support adding a new inductor backend using PrivateUse1 

### LINKED ISSUE #129952: Enhance the compiler module to extend  a new backend using PrivateUse1
### 🚀 The feature, motivation and pitch

Since version 2.1, PyTorch has fully supported the extension of new backends using PrivateUse1 and has provided a detailed user guide. The documentation is as follows:[Extending dispatcher for a new backend in C++](https://pytorch.org/tutorials/advanced/extend_dispatcher.html)
However, in Python modules such as torchinductor and torchdynamo, there are still some areas that need improvement to support new devices using PrivateUse1.
For example, within this init_backend_registration() function: https://github.com/pytorch/pytorch/blob/main/torch/_inductor/codegen/common.py#L194
```
@functools.lru_cache(None)
def init_backend_registration():
    from .cpp import CppScheduling
    from .cpp_wrapper_cpu import CppWrapperCpu
    from .cpp_wrapper_cuda import CppWrapperCuda
    from .cuda_combined_scheduling import CUDACombinedScheduling
    from .halide import HalideScheduling
    from .triton import TritonScheduling
    from .wrapper import WrapperCodeGen

    if get_scheduling_for_device("cpu") is None:
        cpu_backends = {"cpp": CppScheduling, "halide": HalideScheduling}
        register_backend_for_device(
            "cpu",
            lambda *args, **kwargs: cpu_backends[config.cpu_backend](*args, **kwargs),
            WrapperCodeGen,
            CppWrapperCpu,
        )

    if get_scheduling_for_device("cuda") is None:
        # CUDACombinedScheduling combines Triton and CUDA C++ scheduling for CUDA devices via delegation
        cuda_backends = {"triton": CUDACombinedScheduling, "halide": HalideScheduling}
        register_backend_for_device(
            "cuda",
            lambda *args, **kwargs: cuda_backends[config.cuda_backend](*args, **kwargs),
            WrapperCodeGen,
            CppWrapperCuda,
        )

    if get_scheduling_for_device("xpu") is None:
        register_backend_for_device("xpu", TritonScheduling, WrapperCodeGen)
```
In the above function implementation, registration is only done for the CPU, CUDA, and XPU backends, lacking registration for PrivateUse1 devices.
When enhancing this function, in addition to adding support for the PrivateUse1 backend, it's necessary to also support passing in custom TritonScheduling and WrapperCodegen parameters.

### Alternatives

A simple implementation might look like this:
```
@functools.lru_cache(None)
def init_backend_registration():
    from .cpp import CppScheduling
    from .cpp_wrapper_cpu import CppWrapperCpu
    from .cpp_wrapper_cuda import CppWrapperCuda
    from .cuda_combined_scheduling import CUDACombinedScheduling
    from .halide import HalideScheduling
    from .triton import TritonScheduling
    from .wrapper import WrapperCodeGen

    if get_scheduling_for_device("cpu") is None:
        cpu_backends = {"cpp": CppScheduling, "halide": HalideScheduling}
        register_backend_for_device(
            "cpu",
            lambda *args, **kwargs: cpu_backends[config.cpu_backend](*args, **kwargs),
            WrapperCodeGen,
            CppWrapperCpu,
        )

    if get_scheduling_for_device("cuda") is None:
        # CUDACombinedScheduling combines Triton and CUDA C++ scheduling for CUDA devices via delegation
        cuda_backends = {"triton": CUDACombinedScheduling, "halide": HalideScheduling}
        register_backend_for_device(
            "cuda",
            lambda *args, **kwargs: cuda_backends[config.cuda_backend](*args, **kwargs),
            WrapperCodeGen,
            CppWrapperCuda,
        )

    if get_scheduling_for_device("xpu") is None:
        register_backend_for_device("xpu", TritonScheduling, WrapperCodeGen)
    
    private_backend = torch._C._get_privateuse1_backend_name()                                                                                                                                                        
    if private_backend != "privateuseone":
        if get_scheduling_for_device(private_backend) is None:
            from torch.utils.backend_registration import _get_custom_mod_func
            try:    
                triton_scheduling = _get_custom_mod_func("TritonScheduling")
                wrapper_codegen = _get_custom_mod_func("WrapperCodeGen")
                if triton_scheduling and wrapper_codegen:
                    register_backend_for_device(private_backend, triton_scheduling, wrapper_codegen)
            except RuntimeError:
                pass    
```

### Additional context

For users extending backends using PrivateUse1, they only need to add a custom TritonScheduling and WrapperCodegen within their extension package. These can either inherit from the native classes or directly return the native classes.

cc @ezyang @anijain2305 @chauhang @penguinwu @NmomoN @mengpenghui @fwenguang @cdzhan @1274085042 @PHLens

### PR DESCRIPTION
**Pull request by peaceorwell**
Add handling custom device registered by PrivateUse1 in init_backend_registration() func

Fixes #129952

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

**Comment by ezyang** [comment]
I don't want to apply too much stop energy here, because the stakes are low, but it is very unclear to me why *this* particular initialization code is the correct generic initialization code for private use backend. Perhaps it would be more clearly correct if you show in the cpp extension tests for privateuse1 how it should be used with PT2 as well.

In particular, my problem is that what does it even mean to add an Inductor backend for PrivateUse1? This is not really a concept we have at all.

File: torch/_inductor/codegen/common.py [hunk 1]
```diff
@@ -222,6 +222,18 @@ def init_backend_registration():
...
 
+    private_backend = torch._C._get_privateuse1_backend_name()
+    if private_backend != "privateuseone":
+        if get_scheduling_for_device(private_backend) is None:
+            from torch.utils.backend_registration import _get_custom_mod_func
+            try:
+                triton_scheduling = _get_custom_mod_func("TritonScheduling")
+                wrapper_codegen = _get_custom_mod_func("WrapperCodeGen")
+                if triton_scheduling and wrapper_codegen:
+                    register_backend_for_device(private_backend, triton_scheduling, wrapper_codegen)
```
Comment (jansel) [inline torch/_inductor/codegen/common.py:233]: You can call this function from outside PyTorch tree as well.

Comment (peaceorwell) [reply torch/_inductor/codegen/common.py:233]: @jansel Thanks for your comments.
When calling this function out-of-tree, some profiler testcases will fail.
You can reproduce it using the following method:
Add below codes at： https://github.com/pytorch/pytorch/blob/main/test/profiler/test_profiler.py#L347
```
        from torch._inductor.codegen.common import register_backend_for_device                                                                                                                                    
        from torch._inductor.codegen.triton import TritonScheduling
        from torch._inductor.codegen.wrapper import WrapperCodeGen
             
        register_backend_for_device("apu", TritonScheduling, WrapperCodeGen)
```
and then execute `python test/profiler/test_profiler.py -k TestProfiler.test_source_multithreaded_basic_work_in_main_thread_False`.
The following error will be encountered:
```
======================================================================
FAIL: test_source_multithreaded_basic_work_in_main_thread_False (__main__.TestProfiler)
Test various threading configurations.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/zhouyusong/venv_2.4/lib/python3.10/site-packages/torch/testing/_internal/common_utils.py", line 2739, in wrapper
    method(*args, **kwargs)
  File "/home/zhouyusong/venv_2.4/lib/python3.10/site-packages/torch/testing/_internal/common_utils.py", line 446, in instantiated_test
    test(self, **param_kwargs)
  File "/home/zhouyusong/pytorch_2.4/test/profiler/test_profiler.py", line 459, in test_source_multithreaded
    self.assertEqual(
  File "/home/zhouyusong/venv_2.4/lib/python3.10/site-packages/torch/testing/_internal/common_utils.py", line 3631, in assertEqual
    raise error_metas.pop()[0].to_error(
AssertionError: Scalars are not equal!

Expected 2 but got 3.
Absolute difference: 1
Relative difference: 0.5
2, Counter({2: 96, 1: 63, 18446744073709551615: 5})
```
It appears that using the out-of-tree method prematurely triggers the import of some Python modules.
I am still trying to locate and resolve the issue.

**Comment by peaceorwell** [comment]
> I don't want to apply too much stop energy here, because the stakes are low, but it is very unclear to me why _this_ particular initialization code is the correct generic initialization code for private use backend. Perhaps it would be more clearly correct if you show in the cpp extension tests for privateuse1 how it should be used with PT2 as well.
> 
> In particular, my problem is that what does it even mean to add an Inductor backend for PrivateUse1? This is not really a concept we have at all.

Before answering your question, allow me to describe the problem I've encountered. 
In [this document](https://pytorch.org/get-started/pytorch-2.0/#developervendor-experience), PyTorch provides three methods for extending the compiler backend. We want to follow the second method (Inductor backend) for backend extension. To add new backend registration for the inductor backend, we tried two approaches:

1. Calling register_backend_for_device in our extension package (out-of-tree) to add it.
2. Using monkey patching to override the init_backend_registration() function.

Both methods work successfully. However, they both cause an error in test_source_multithreaded within profiler/test_profilter.py. The reason is that both of these out-of-tree methods lead to the premature initialization of the AsyncCompile process pool. When we set TORCH_WARM_POOL=0, these test cases can pass. 
The changes in this PR are just an optional path to bypass this issue, and do not indicate a problem with the PrivateUse1 mechanism itself.

**Review by jansel — APPROVED** [review]

**Pull request closed without merging**